### PR TITLE
S3 Lock service

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/6827

Problem Summary: We need one and only one tiflash node run S3 lock service.

### What is changed and how it works?

Depends on
- OwnerManager: https://github.com/pingcap/tiflash/pull/6863
- diff [1e7378...b918ba5](https://github.com/pingcap/tiflash/pull/6874/files/1db1d48dc0708f00884ed629f87e52d9e1d402d1..feeb2d4193ed80b588e8de53f8b2d0e4c182d6bd)
- kvproto: https://github.com/pingcap/kvproto/pull/1060
- https://github.com/pingcap/tiflash/pull/6878

#### S3Filename
Generate the S3 object keys for local file or parsing the key from S3 object key

#### S3LockService
S3LockService provides two RPC interface and ensure these two operations with the same key (the S3 key of CheckpointDataFile/DTFile)  is atomic on S3
- TryAddLock
- TryMarkDelete

If the owner changed, this service will return NotOwner with the new owner id to client. Client should retry the request to new owner.

#### S3LockClient
Make rpc request to the S3 GC owner. S3LockClient will automatically retry when owner is changed

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Disaggregated/S3LockService.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+#include <unordered_map>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#pragma GCC diagnostic ignored "-Wnon-virtual-dtor"
+#ifdef __clang__
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
+#include <kvproto/disaggregated.pb.h>
+#include <kvproto/kvrpcpb.pb.h>
```
Comment (Lloyd-Pottiger) [inline dbms/src/Flash/Disaggregated/S3LockService.h:36]: maybe useless header?

File: dbms/src/Storages/S3/S3Filename.cpp [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+#include <Common/Exception.h>
+#include <Common/StringUtils/StringRefUtils.h>
+#include <Storages/S3/S3Filename.h>
+#include <re2/re2.h>
+#include <re2/stringpiece.h>
+
+#include <magic_enum.hpp>
+#include <string_view>
+
+#include "Storages/DeltaMerge/DeltaMergeDefines.h"
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/S3/S3Filename.cpp:24]:
```suggestion
#include <Storages/DeltaMerge/DeltaMergeDefines.h>
```

File: dbms/src/Storages/S3/S3Filename.cpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+S3Filename S3Filename::newCheckpointManifest(UInt64 store_id, UInt64 upload_seq)
+{
+    return S3Filename{
+        .type = S3FilenameType::CheckpointManifest,
+        .store_id = store_id,
+        .path = fmt::format("mf_{}", upload_seq),
+    };
+}
+
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/S3/S3Filename.cpp:304]:
```suggestion
```

File: dbms/src/Flash/Disaggregated/S3LockService.h (hunk 1 above)
Comment (JaySon-Huang) [reply dbms/src/Flash/Disaggregated/S3LockService.h]: removed

**Review by Lloyd-Pottiger — APPROVED** [review]

**Review by flowbehappy — APPROVED** [review]

**Pull request merged**